game-core/src/main/java/games/strategy/engine/chat/LobbyChatTransmitter.java

```diff
@@ -0,0 +1,71 @@
+package games.strategy.engine.chat;
+
+import games.strategy.engine.lobby.client.LobbyClient;
+import java.util.Collection;
+import org.triplea.domain.data.PlayerName;
+import org.triplea.http.client.lobby.chat.ChatParticipant;
+import org.triplea.http.client.lobby.chat.LobbyChatClient;
+
+/**
+ * Chat transmitter designed to work with lobby, sends and receives messages over Websocket. This
+ * class provides a send API and does wiring to connect callbacks from a {@code LobbyChatClient} to
+ * a {@code ChatClient}.
+ */
+// TODO: Project#12 test-me
+public class LobbyChatTransmitter implements ChatTransmitter {
+  private final LobbyChatClient lobbyChatClient;
+  private final PlayerName localPlayerName;
+
+  public LobbyChatTransmitter(final LobbyClient lobbyClient) {
+    this.localPlayerName = lobbyClient.getPlayerName();
+    this.lobbyChatClient = lobbyClient.getHttpLobbyClient().getLobbyChatClient();
+  }
+
+  @Override
+  public void setChatClient(final ChatClient chatClient) {
+    lobbyChatClient.addPlayerStatusListener(chatClient::statusUpdated);
+    lobbyChatClient.addPlayerLeftListener(chatClient::participantRemoved);
+    lobbyChatClient.addPlayerJoinedListener(chatClient::participantAdded);
+    lobbyChatClient.addChatMessageListener(chatClient::messageReceived);
+    lobbyChatClient.addConnectedListener(chatClient::connected);
+
+    lobbyChatClient.addPlayerSlappedListener(
+        slapEvent -> {
+          if (slapEvent.getSlapped().equals(localPlayerName)) {
+            chatClient.slappedBy(slapEvent.getSlapper());
+          } else {
+            chatClient.playerSlapped(slapEvent.getSlapper() + " slapped " + slapEvent.getSlapped());
+          }
+        });
```

Comment:
Could it ever be the case that this method is called multiple times so that a `ChatClient` instance must stay in memory for the lifetime of the `lobbyChatClient` even when there are no other references?
I think lobbyChatClient should provide a way to clear out old listeners if a new one is set, or alternatively this method should prevent calling itself multiple times

Reply:
We've essentially a glorified callback, perhaps that perspective will make this make more sense. To another extent, with multiple chat clients, all objects should be going out of scope at the same time. If multiple listeners are interested in chat events, when the chat goes away, so should all the listeners. 

It's not ideal this is even a `set` method, that is largely a constraint of working with the existing chat code. It does not also make sense to have singleton instances as un-registering other listeners would be a surprising side effect as well. The dependency is a bit on its head overall, as we would want the listeners to be constructor args to `lobbyChatClient`, but that creates a circular dependency.

Reply:
I'd agree that it's problematic that this looks like a setter from the outside.
If this method would rather be an add* method I probably wouldn't even have bothered to mention this, because in this case nobody expects this method to replace any potential previous entry.